filebeat/input/filestream/internal/input-logfile/store.go

```diff
@@ -211,6 +211,51 @@ func (s *sourceStore) CleanIf(pred func(v Value) bool) {
 	}
 }
 
+// FixUpIdentifiers copies an existing resource to a new ID and marks the previous one
+// for removal. This method does not check whether the resource belongs to this store
+// beause it is meant to migrate inputs without IDs to inputs with IDs without
+//  duplicating data.
+func (s *sourceStore) FixUpIdentifiers(getNewID func(v Value) (string, interface{})) {
+	s.store.ephemeralStore.mu.Lock()
+	defer s.store.ephemeralStore.mu.Unlock()
+
+	for key, res := range s.store.ephemeralStore.table {
+		if !s.identifier.MatchesInput(key) {
+			continue
+		}
+
+		res.lock.Lock()
+
+		newKey, updatedMeta := getNewID(res)
+		if len(newKey) > 0 && res.internalState.TTL > 0 {
+			if _, ok := s.store.ephemeralStore.table[newKey]; ok {
+				res.lock.Unlock()
+				continue
+			}
+
+			// Pending updates due to events that have not yet been ACKed
+			// are not included in the copy. Collection on
+			// the copy start from the last known ACKed position.
+			// This might lead to duplicates if configurations are adapted
+			// for inputs with the same ID are changed.
```

Comment:
> 			// This might lead to duplicates if configurations are adapted
> 			// for inputs with the same ID are changed.

I'm not sure what this comment is telling me. Can you clarify it?

Reply:
Neither do I :/.
This was a copy and paste from `UpdateIdentifiers`: https://github.com/elastic/beats/blob/4a71e5e7eb8faa711785aee434a2aadd248439e0/filebeat/input/filestream/internal/input-logfile/store.go#L236-L244

I could not use `UpdateIdentifiers` for this feature because it wasn't updating the in-memory registry (`s.store.ephemeralStore`), so I copied it and did the small modifications I needed.

Reply:
@kvch could you help us out here?

Reply:
The last sentence does not make much sense. :facepalm: :rofl: 

What I wanted to explain is just a corner case, it does not impact us because we do not update identifiers during data collection at the moment.

However, if identifiers are updated during data collection, it is possible that there are state updates that haven't been ACKed (by the output) when the `copyWithNewKey` is called. `copyWithNewKey` copies the existing state information with a new ID. So if there are outstanding updates to the state with the old ID, those updates will not be persisted to the newly created state with the new ID. So when Filebeat picks up the state with the new ID for the same source, the data will be duplicated because the updates of the old state are "lost".

Example for data duplication: Filebeat is collecting a `test.log` file with the ID of `filestream::test.log` and the last ACKed offset is 23. If the file is being harvested and some events are not yet been ACKed by Elasticsearch until the offset of 50. If Filebeat updates the ID of the state from `filestream::test.log` to `filestream::1234` the offset will be 23 in the registry. The data from bytes 23 to 50 in the file `test.log` is going to be resent to Elasticsearch regardless of the previous ACK because the ACK did not make it to Filebeat in time.

But as I mentioned above at the moment we do not update identifiers during data collection.